### ISSUE TITLE
question about get_rlloss()

### ISSUE THREAD

**Issue by Miracle1991** [hustvl/DiffusionDriveV2#8]
`

def get_rlloss(self, ego_query,agents_query,bev_feature,bev_spatial_shape,status_encoding,status_feature,camera_feature, targets,global_img, eta, old_pred):

        old_diffusion_output = old_pred['all_diffusion_output']
        advantages = old_pred['advantages']

        chains = old_diffusion_output[...,:-1]
        chains_prev = old_diffusion_output[...,1:]  

        step_num = 10
        bs = chains.shape[0]
        device = chains.device
        self.diffusionrl_scheduler.set_timesteps(1000, device)
        step_ratio = 20 / step_num
        roll_timesteps = (np.arange(0, step_num) * step_ratio).round()[::-1].copy().astype(np.int64)
        roll_timesteps = torch.from_numpy(roll_timesteps).to(device)

        all_log_probs = []
        poses_reg_steps_list = []
        poses_cls_steps_list = []
        for i, k in enumerate(roll_timesteps[:]):
            diffusion_output = chains[..., i]
            ego_fut_mode = diffusion_output.shape[1]
            x_boxes = torch.clamp(diffusion_output, min=-1, max=1)
            noisy_traj_points = self.denorm_odo(x_boxes)

            # 2. proj noisy_traj_points to the query
            traj_pos_embed = gen_sineembed_for_position(noisy_traj_points,hidden_dim=64)
            traj_pos_embed = traj_pos_embed.flatten(-2)
            traj_feature = self.plan_anchor_encoder(traj_pos_embed)
            traj_feature = traj_feature.view(bs,ego_fut_mode,-1)

            timesteps = k
            if not torch.is_tensor(timesteps):
                # TODO: this requires sync between CPU and GPU. So try to pass timesteps as tensors if you can
                timesteps = torch.tensor([timesteps], dtype=torch.long, device=diffusion_output.device)
            elif torch.is_tensor(timesteps) and len(timesteps.shape) == 0:
                timesteps = timesteps[None].to(diffusion_output.device)
            
            # 3. embed the timesteps
            timesteps = timesteps.expand(diffusion_output.shape[0])
            time_embed = self.time_mlp(timesteps)
            time_embed = time_embed.view(bs,1,-1)

            # 4. begin the stacked decoder
            poses_reg_list, poses_cls_list = self.diff_decoder(
                traj_feature, noisy_traj_points, bev_feature, bev_spatial_shape,
                agents_query,
                ego_query,
                time_embed, status_encoding, global_img
            )
            poses_reg_steps_list.append(poses_reg_list)
            poses_cls_steps_list.append(poses_cls_list)
            poses_reg = poses_reg_list[-1]
            poses_cls = poses_cls_list[-1]
            x_start = poses_reg[..., :2] # bs G*N 8 2
            # x_start = poses_reg
            x_start = self.norm_odo(x_start)
            _, log_prob, _ = self.diffusionrl_scheduler.step(
                model_output=x_start,
                timestep=k,
                sample=diffusion_output,
                eta=eta,
                prev_sample=chains_prev[...,i]
            )
            all_log_probs.append(log_prob)
        all_log_probs = torch.stack(all_log_probs, dim=-1) # BG N step_num
        per_token_logps = all_log_probs.view(bs, self.num_groups * self.ego_fut_mode, -1)  # B G*N step_num

        per_token_loss = -torch.exp(per_token_logps - per_token_logps.detach()) * advantages

        # ---------- (1) RL 损失，保留 batch 维 ----------
        mask_nz    = per_token_loss != 0            # (B,G,T)
        RL_loss_b  = (per_token_loss * mask_nz).sum(dim=1) \
                    / mask_nz.sum(dim=1).clamp_min(1)      # (B,T)
        RL_loss_b = RL_loss_b.mean(dim=-1)  # (B,)
        # ---------- (2) IL 损失，先算每个 batch ----------
        IL_loss_b = torch.zeros_like(RL_loss_b)     # (B,)
        target_traj = targets['trajectory'].unsqueeze(1).repeat(1, self.num_groups*self.ego_fut_mode, 1, 1)
        
        for poses_reg_list in poses_reg_steps_list:                 # 5 个 time-step
            for poses_reg in poses_reg_list:                        # 2 层 decoder
                traj_l1 = F.l1_loss(poses_reg[...,:2], target_traj[...,:2], reduction='none')  # (B,G,T,C)
                IL_loss_b += traj_l1.mean()                   # 加到 (B,)

        IL_loss_b /= (len(poses_reg_steps_list) * len(poses_reg_steps_list[0]))  # 取平均
        has_positive   = (advantages > 0).any(dim=2).any(dim=1)         # (B,) bool

        il_weight  = torch.where(has_positive == 0,          # (B,)
                                torch.tensor(1.0,  device=RL_loss_b.device),
                                torch.tensor(0.1, device=RL_loss_b.device))
        loss_b     = RL_loss_b + il_weight*IL_loss_b  # (B,)
        loss       = loss_b.mean()                        # 标量
        return {"loss": loss}`

在这一句 per_token_loss = -torch.exp(per_token_logps - per_token_logps.detach()) * advantages
我比较疑惑这个操作，-torch.exp(per_token_logps - per_token_logps.detach())应该一直是-1，所以计算得到per_token_logps的过程都是没用的？

**Comment by Doctor-James** [comment]
这是GRPO的一种简化版本，per_token_logps - per_token_logps.detach()虽然数值是0，但是可以反传梯度，因为我们没有old model，相当于设置重要性采样系数为1。这种简化形式在各大GRPO的开源代码中都有使用，比如[R1-V](https://github.com/StarsfieldAI/R1-V/blob/e35f97e566357631a2602accdd79610b752d8f0b/src/r1-v/src/open_r1/trainer/grpo_trainer.py#L465)
